runtime/src/bank.rs

```diff
@@ -12525,101 +12525,133 @@ pub(crate) mod tests {
+        solana_logger::setup();
+        do_test_clean_dropped_unrooted_banks(FreezeBank1::Yes);
+    }
+
+    #[test]
+    fn test_clean_dropped_unrooted_unfrozen_banks() {
+        solana_logger::setup();
+        do_test_clean_dropped_unrooted_banks(FreezeBank1::No);
+    }
+
+    /// A simple enum to toggle freezing Bank1 or not.  Used in the clean_dropped_unrooted tests.
+    enum FreezeBank1 {
+        No,
+        Yes,
+    }
+
+    fn do_test_clean_dropped_unrooted_banks(freeze_bank1: FreezeBank1) {
+        //! Test that dropped unrooted banks are cleaned up properly
         //!
         //! slot 0:       bank0 (rooted)
         //!               /   \
-        //! slot 1:      /   bank1 (unrooted, frozen, and dropped)
+        //! slot 1:      /   bank1 (unrooted and dropped)
         //!             /
         //! slot 2:  bank2 (rooted)
         //!
         //! In the scenario above, when `clean_accounts()` is called on bank2, the keys that exist
         //! _only_ in bank1 should be cleaned up, since those keys are unreachable.
-        //
-        solana_logger::setup();
+        //!
+        //! The following scenarios are tested:
+        //!
+        //! 1. A key is written _only_ in an unrooted bank (key1)
+        //!     - In this case, key1 should be cleaned up
+        //! 2. A key is written in both an unrooted _and_ rooted bank (key3)
+        //!     - In this case, key3's ref-count should be decremented correctly
+        //! 3. A key with zero lamports is _only_ in an unrooted bank (key4)
+        //!     - In this case, key4 should be cleaned up
+        //! 4. A key with zero lamports is in both an unrooted _and_ rooted bank (key5)
+        //!     - In this case, key5's ref-count should be decremented correctly
 
         let (genesis_config, mint_keypair) = create_genesis_config(100);
         let bank0 = Arc::new(Bank::new(&genesis_config));
 
         let collector = Pubkey::new_unique();
-        let pubkey1 = Pubkey::new_unique();
-        let pubkey2 = Pubkey::new_unique();
 
-        bank0.transfer(2, &mint_keypair, &pubkey2).unwrap();
+        let key1 = Keypair::new(); // only touched in bank1
+        let key2 = Keypair::new(); // only touched in bank2
+        let key3 = Keypair::new(); // touched in both bank1 and bank2
+        let key4 = Keypair::new(); // in only bank1, and has zero lamports
+        let key5 = Keypair::new(); // in both bank1 and bank2, and has zero lamports
+
+        bank0.transfer(2, &mint_keypair, &key2.pubkey()).unwrap();
         bank0.freeze();
 
         let slot = 1;
         let bank1 = Bank::new_from_parent(&bank0, &collector, slot);
-        bank1.transfer(3, &mint_keypair, &pubkey1).unwrap();
-        bank1.freeze();
+        bank1.transfer(3, &mint_keypair, &key1.pubkey()).unwrap();
+
+        // make key4 have zero lamports
+        bank1.transfer(8, &mint_keypair, &key4.pubkey()).unwrap();
+        bank1.transfer(8, &key4, &key3.pubkey()).unwrap();
```

Comment:
heh, creative. Here's also a more straightforward way to store a zero lamport account https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L6583-L6587

Reply:
OK, I've updated the tests. I've used `Bank::store_account()` instead of `Bank::transfer()`. I'm assuming it's OK to use this version instead of `AccountsDb::store_uncached()`. Since the tests are in `bank.rs`, I wanted to only use `Bank` functions, if possible.